File: src/y12.js
console.log(function() {

  // Fixpoints
  // Higher Order Functions
  // Functional Refactorings
  // (1) Tennent Correspondence Principle
  // (2) Introduce Binding
  // (3) Rebinding
  // (4) Function Wrapping
  // (5) Inline Function

  error = function(n) { throw Error("SHOULD NOT CALL") }

  improver = function(partial) {
    return function(n) {
      return n==0 ? 1 : n * partial(n-1)
    }
  }

  y = function(f) {
    return function(g) { return function(n) { return f(g(g))(n) } }(
           function(g) { return function(n) { return f(g(g))(n) } }
    )
  }

  fx = y(improver)

  return fx(5)

}())

File: src/y12.clj
(println
 (do

   (def error (fn [x] (throw (Throwable. "SHOULD NOT BE CALLED"))))

   ;; Fixpoints
   ;; Higher Order Functions
   ;; Functional Refactorings
   ;; (1) Tennent Correspondence Principle
   ;; (2) Introduce Binding
   ;; (3) Rebind
   ;; (4) Function Wrapping
   ;; (5) Function Inlining

   (def y
        (fn [improver]
          ((fn [gen] (gen gen))
           (fn [gen] (fn [n] ((improver (gen gen)) n))))))

   (def fact-improver
        (fn [partial]
          (fn [n] (if (zero? n) 1 (* n (partial (- n 1)))))))

   (def fact (y fact-improver))

   ;; fact == (fact-improver fact)

   (fact 5)

   ))
